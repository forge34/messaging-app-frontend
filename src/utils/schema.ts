export interface User {
  pk: number;
  id: string;
  name: string;
  password: string;
  messages: Message[];
  conversations: Conversation[];

  relatedToCurrent: boolean;
}

export interface Message {
  id: string;
  body: string;
  createdAt: Date;
  authorId: string;
  conversationId: string;
  author: User;
  Conversation: Conversation;
  ownMessage:boolean
}

export interface Conversation {
  id: string;
  title: string;
  messages: Message[];
  users?: User[];
}
