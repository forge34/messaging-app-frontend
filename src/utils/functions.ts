export const randomIp = () =>
  Math.floor(Math.random() * 255) +
  1 +
  "." +
  Math.floor(Math.random() * 255) +
  "." +
  Math.floor(Math.random() * 255) +
  "." +
  Math.floor(Math.random() * 255);
