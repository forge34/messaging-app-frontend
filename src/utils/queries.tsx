import { QueryClient, queryOptions } from "@tanstack/react-query";
import { Params } from "react-router-dom";
import { Conversation, User } from "./schema";

const getConversationById = (id: string | undefined) =>
  queryOptions({
    queryKey: ["conversations", id],
    queryFn: async (): Promise<Conversation> => {
      const res = await fetch(
        `${import.meta.env.VITE_API}/conversation/${id}`,
        {
          mode: "cors",
          credentials: "include",
        },
      );
      if (res.status === 401) {
        throw res;
      }

      return res.json();
    },
  });

const conversationIdLoader =
  (queryClient: QueryClient) =>
  async ({ params }: { params: Params<"id"> }) => {
    const query = getConversationById(params.id);

    return queryClient.ensureQueryData(query);
  };

const getUserConversations = () =>
  queryOptions({
    queryKey: ["conversations"],
    queryFn: async (): Promise<Array<Conversation>> => {
      const res = await fetch(
        `${import.meta.env.VITE_API}/conversation/currentuser`,
        {
          mode: "cors",
          credentials: "include",
        },
      );
      if (res.status === 401) {
        throw res;
      }

      return res.json();
    },
  });

const conversationLoader = (queryClient: QueryClient) => async () => {
  const query = getUserConversations();

  return queryClient.ensureQueryData(query);
};

const getUsers = () =>
  queryOptions({
    queryKey: ["users"],
    queryFn: async (): Promise<Array<User>> => {
      const res = await fetch(`${import.meta.env.VITE_API}/users`, {
        method: "get",
        mode: "cors",
        credentials: "include",
      });

      if (res.status === 401) {
        throw res;
      }

      return res.json();
    },
    staleTime: 1000 * 60 * 2,
  });

const userLoader = (queryClient: QueryClient) => async () => {
  const query = getUsers();

  return queryClient.ensureQueryData(query);
};
export {
  userLoader,
  getUsers,
  getUserConversations,
  conversationLoader,
  getConversationById,
  conversationIdLoader,
};
