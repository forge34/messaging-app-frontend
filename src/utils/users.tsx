import { QueryClient, queryOptions } from "@tanstack/react-query";
import { Conversation, User } from "./schema";

const getUserConversations = () =>
  queryOptions({
    queryKey: ["user-conversations"],
    queryFn: async (): Promise<Array<Conversation>> => {
      const res = await fetch(
        `${import.meta.env.VITE_API}/conversation/currentuser`,
        {
          mode: "cors",
          credentials: "include",
        },
      );

      return res.json();
    },
  });

const conversationLoader = (queryClient: QueryClient) => async () => {
  const query = getUserConversations();

  return queryClient.ensureQueryData(query);
};

const getUsers = () =>
  queryOptions({
    queryKey: ["users"],
    queryFn: async (): Promise<Array<User>> => {
      const res = await fetch(`${import.meta.env.VITE_API}/users`, {
        method: "get",
        mode: "cors",
        credentials: "include",
      });

      if (!res.ok) {
        throw new Error("error during fetch");
      }

      return res.json();
    },
    staleTime: 1000 * 60 * 2,
  });

const userLoader = (queryClient: QueryClient) => async () => {
  const query = getUsers();

  return queryClient.ensureQueryData(query);
};
export { userLoader, getUsers, getUserConversations, conversationLoader };
