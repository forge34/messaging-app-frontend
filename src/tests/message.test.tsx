import { render, screen } from "@testing-library/react";
import Message, { MessageProps } from "../components/message";
import { ConversationSchema, MessageSchema, UserSchema } from "../utils/schema";

const testUser: UserSchema = {
  pk: 1,
  id: "cm0xul6bz000109mb8i2t3j1m",
  name: "Test User",
  imgUrl: "url.com",
  conversations: [] as ConversationSchema[],
  messages: [] as MessageSchema[],
  relatedToCurrent: false,
};

const testMessage: MessageProps = {
  author: testUser,
  body: "Hello",
  ownMessage: true,
  id: "cm0xulkra000209mb6b6nbwp1",
};

describe("message component test", () => {
  test("should match snapshot", () => {
    const { container } = render(<Message {...testMessage}></Message>);

    expect(container).toMatchSnapshot();
  });

  test("should render correct message text", () => {
    render(<Message {...testMessage}></Message>);

    expect(screen.getByText("Hello")).toBeInTheDocument();
  });
});
