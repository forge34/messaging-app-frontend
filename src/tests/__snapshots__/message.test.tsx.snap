// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`message component test > should match snapshot 1`] = `
<div>
  <div
    class="message own"
    id="cm0xulkra000209mb6b6nbwp1"
  >
    <img
      height="32"
      src="url.com"
      width="32"
    />
    <p>
      Hello
    </p>
    <img
      src=""
    />
  </div>
</div>
`;
